mempool/mempool.go

```diff
@@ -0,0 +1,362 @@
+		}
+	}
+
+	newNode := &memPoolTxn{Txn: *userTxn, Next: nil}
+	p.memTxnList.push(newNode)
+
+	select {
+	case p.txPushed <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+func (p *Pool) validate(userTxn *BroadcastedTransaction) error {
+	if p.memTxnList.len+1 >= p.maxNumTxns {
+		return ErrTxnPoolFull
+	}
+
+	switch t := userTxn.Transaction.(type) {
+	case *core.DeployTransaction:
+		return fmt.Errorf("deploy transactions are not supported")
+	case *core.DeployAccountTransaction:
+		if !t.Nonce.IsZero() {
+			return fmt.Errorf("validation failed, received non-zero nonce %s", t.Nonce)
+		}
+	case *core.DeclareTransaction:
+		nonce, err := p.state.ContractNonce(t.SenderAddress)
+		if err != nil {
+			return fmt.Errorf("validation failed, error when retrieving nonce, %v", err)
+		}
+		if nonce.Cmp(t.Nonce) > 0 {
+			return fmt.Errorf("validation failed, existing nonce %s, but received nonce %s", nonce, t.Nonce)
+		}
+	case *core.InvokeTransaction:
+		if t.TxVersion().Is(0) { // cant verify nonce since SenderAddress was only added in v1
+			return fmt.Errorf("invoke v0 transactions not supported")
+		}
+		nonce, err := p.state.ContractNonce(t.SenderAddress)
+		if err != nil {
+			return fmt.Errorf("validation failed, error when retrieving nonce, %v", err)
+		}
+		if nonce.Cmp(t.Nonce) > 0 {
+			return fmt.Errorf("validation failed, existing nonce %s, but received nonce %s", nonce, t.Nonce)
+		}
+	case *core.L1HandlerTransaction:
+		// todo: verification of the L1 handler nonce requires checking the
+		// message nonce on the L1 Core Contract.
+	}
+	return nil
+}
+
+// Pop returns the transaction with the highest priority from the in-memory pool
+func (p *Pool) Pop() (BroadcastedTransaction, error) {
+	return p.memTxnList.pop()
+}
+
+// Remove removes a set of transactions from the pool
+// todo: should be called by the builder to remove txns from the db everytime a new block is stored.
+// todo: in the consensus+p2p world, the txns should also be removed from the in-memory pool.
+func (p *Pool) Remove(hash ...*felt.Felt) error {
+	return errors.New("not implemented")
+}
+
+// Len returns the number of transactions in the in-memory pool
+func (p *Pool) Len() int {
+	return p.memTxnList.len
+}
+
+// Len returns the number of transactions in the persistent pool
+func (p *Pool) LenDB() (int, error) {
+	p.wg.Add(1)
+	defer p.wg.Done()
+	txn, err := p.db.NewTransaction(false)
+	if err != nil {
+		return 0, err
+	}
+	lenDB, err := p.lenDB(txn)
+	if err != nil {
+		return 0, err
+	}
+	return lenDB, txn.Discard()
+}
+
+func (p *Pool) lenDB(txn db.Transaction) (int, error) {
+	var l int
+	err := txn.Get(db.MempoolLength.Key(), func(b []byte) error {
+		l = int(new(big.Int).SetBytes(b).Int64())
+		return nil
+	})
+
+	if err != nil && errors.Is(err, db.ErrKeyNotFound) {
+		return 0, nil
+	}
+	return l, err
+}
+
+func (p *Pool) updateLen(txn db.Transaction, l int) error {
+	return txn.Set(db.MempoolLength.Key(), new(big.Int).SetInt64(int64(l)).Bytes())
+}
```

Comment:
This method is used only once, I think we can safely inline it inside `writeToDB` since it is not used anywhere else.